docs/CIP_6_Pipelines_Registry.md

```diff
@@ -0,0 +1,309 @@
+
+#### Future work: Parallel pipeline execution
+Not all pipeline stages need to executed sequentially. In the future, we can add support for parallel pipeline execution by changing how the pipeline is defined.
+
+## Storage Pipelines
+Chroma right now implicitly stores your documents, however we also want to support multimodal embeddings and part of this CIP is to propose we deprecate documents= in favor of data=. The contract then becomes that the data you input (or that is generated by your pipeline) is stored by default and can be retrieved by the user in some form. We also want to support storing mixed modalities in a collection. For example, a CLIP based collection should be able
+to store both text and images.
+
+We propose to add Storage Pipeline steps capable of storing the inputted data into a Storage layer, and then storing a string reference to that data in the metadata of the document. This is already how the document is currently stored and backed by the metadata segment. The metadata segment stores a KV pair "chroma:document" -> the document contents. For example, say you had an image in memory and wanted to store it in Chroma, you could do something like this:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(ImageStoragePipelineStep()))
+collection.add(data=[image], ids=[0]) # image is a PIL image, or numpy array etc
+
+# The ImageStoragePipelineStep would store the image in the storage layer, and then pass through the url to the image instead.
+# The metadata segment would then store the KV pair "chroma:document" -> "https://storage.com/image_1"
+```
+
+We can ship two StoragePipelines to start
+
+1. LocalStoragePipelineStep - Stores the data in a local file system
+2. S3StoragePipelineStep - Stores the data in S3
+
+For now, we would just return the stored string to the user, since they can decide if they want to load it or not. In the future, we could add a LoadFromStorage output pipeline step that would load the data from the storage layer and return it to the user.
+
+For example, if you wanted to load the data from the storage layer and return it to the user, you could do something like this:
+
+```python
+collection = client.create_collection("my_collection", pipeline=DefaultEmbeddingFunctionPipelineStep(ImageStoragePipelineStep()), query_pipeline=DefaultEmbeddingFunctionPipelineStep(), output_pipeline=LoadFromStorage())
+collection.add(data=[image], ids=[0]) # image is a PIL image, or numpy array etc
+collection.query(data=[image], n_results=100) # LoadFromStorage will run before the results are returned and will load the data from the storage layer and return it to the user in memory
+```
+
+#### Future work: More Storage Pipelines
```

Comment:
another idea I had is this... 

perhaps we should embrace the idea of a "repo per database" or "repo with /mapped folders as DB"

then custom ops can be defined in that github repo. 

Chroma can link a database to a repo and then enable CD - so when you update the repo (on a set branch)- it updates that database. (could also extend to have ideas of dev/staging branches) 

Then we could have a `chroma init` command in the CLI that would set up the scaffold for this repo to make it easy. 

It's a way to use Git to manage the lifecycle of the DB itself.

Reply:
There are three concepts being intermingled here.

1. What is the schema of your chroma instance.
2. How is that schema stored and defined.
3. How is that schema updated.

I don't think OSS Chroma should become a gitops platform, which is what you are suggesting, that feels far out of scope and very specific to one way of doing things. Not all organizations will want that.

We certainly can allow you to declaratively define the schema of your DB in code, and then allow programmatic updates to that schema. This starts to feel a lot like reinventing SQL, but thats a neither here nor there.

I can see hosted chroma offering such functionality, but seems far off.

Reply:
definitely far off and very hypothetical